### PR TITLE
The Reconciler

### PR DESCRIPTION
**Pull request by theKashey**
This implements idea from #703 
Fixes: #650 (HoC), #666 (HoC), and any other HoC or decorator issues.

Idea - instead of relying on registering the variables - relay on react-tree.

Algorithm is simple
 - use something like react-deep-force-update to hydrate the render tree
 - `render` the tree back, swapping the old components, and the new ones on the fly.

Results
The good one - ANY component could be Hot-Reloaded. You can use complex HOC or decorators and still have RHL working🏆 
The bad one - all components should be wrapped by proxy. This actually does not affect performance at all, as long stand-in operates on pure js prototypical inheritance.

To activate reconciler - add `reconciler` prop to AppContainer.

PS: Not to be merged yet.

### CODE REVIEW

File: packages/react-hot-loader/src/patch.dev.js [hunk 1]
```diff
@@ -1,77 +1,12 @@
...
-}
-
-let proxiesByID
-let didWarnAboutID
-let hasCreatedElementsByType
-let idsByType
-let firstInstances
-let knownSignatures
-let didUpdateProxy
+import {
```
Comment (theKashey) [inline packages/react-hot-loader/src/patch.dev.js]: patch.dev was completely rewritten. Things like error reporting about failed registrations or failed replacement were removed, as long they are obsolete with reconciler.

**Comment by neoziro** [comment]
Very good work! The idea is great! I think you have resolved the major issue of this project 😄.

File: packages/react-hot-loader/src/reconciler/ComponentMap.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
+export default class ComponentMap {
+  constructor(useWeakMap) {
+    if (useWeakMap) {
+      this.wm = new WeakMap()
```
Comment (neoziro) [inline packages/react-hot-loader/src/reconciler/ComponentMap.js]: We could rely on WeakMap since we are targetting only modern browsers.

File: packages/react-hot-loader/src/AppContainer.dev.js [hunk 3]
```diff
@@ -2,16 +2,20 @@ import React, { Component } from 'react'
 import PropTypes from 'prop-types'
 import deepForceUpdate from 'react-deep-force-update'
 import { getGeneration } from './updateCounter'
+import hydrate from './reconciler/traverse'
```
Comment (neoziro) [inline packages/react-hot-loader/src/AppContainer.dev.js]: hydrate / traverse

File: packages/react-hot-loader/src/reconciler/deepForceTraverse.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      }
+      return {
+        ...b[index],
+        type: child.type,
+      }
+    }),
+  }
+}
+
+const deepForceTraverse = (instance, stack) => {
```
Comment (neoziro) [inline packages/react-hot-loader/src/reconciler/deepForceTraverse.js]: The goal of this function is to traverse a tree (stack) and swap all components is it right? I think the naming is not good.

File: packages/react-hot-loader/src/reconciler/traverse.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    do {
+      const childStack = {}
+      traverseTree(child, childStack)
+      stack.children.push(childStack)
+      child = child.sibling
+    } while (child)
+  }
+}
+
+export default function hydrate(instance) {
```
Comment (neoziro) [inline packages/react-hot-loader/src/reconciler/traverse.js]: The goal is to traverse React tree and build an AST of it in order to traverse it and swap all swappable components, yes?

**Comment by theKashey** [comment]
Ideally this should be moved into react-mount, as a user defined class comparison function. And, as result we could even remove proxy, as long there will be no need to fake types. 
A single line change in react internals can make things must simpler.
How to sell this idea to the React team?

**Comment by neoziro** [comment]
@gaearon can help us!

**Comment by neoziro** [comment]
@theKashey we can't remove proxy because we have to keep state and variables. So even if we have a custom comparison function proxies are still needed.

**Comment by gaearon** [comment]
Please avoid relying on React internal APIs if you can. We don't want to introduce more dependencies than already exist in `deep-force-update`.

I intend to work on officially supporting hot reloading in React in the beginning of next year. So it's best to coordinate our efforts.

**Comment by theKashey** [comment]
@gaearon - does this mean, that we should not proceed with this PR, as long it uses some React internals, or we could proceed, as long it uses the same internals as deep-force-update?

**Comment by gaearon** [comment]
If it uses the same exact internals it’s okay.

**Comment by theKashey** [comment]
All `react internals` were move into a single (and quite small) file. Also refactored logic a bit (just perform dead code elimination).

**Comment by theKashey** [comment]
@neoziro - just added a couple of tests. Coverage is not great, but enough for a start. 
It is much more important to test this solution against some real edge cases, I am not quite aware of.

Waive if you are keen to merge this to the lerna branch.

**Comment by neoziro** [comment]
@theKashey Thanks! I could take a look on tuesday, not before.

**Comment by gaearon** [comment]
I want to note that I still don't really understand what this is doing, so I'd appreciate a more in-depth explanation of why this works, why it is better than the existing approach, and what are the potential downsides and things that could break.

That said I don't want to block you on my understanding :-) If you're sure this is solid and give it a lot of testing in large codebases (I'm sure plenty of people would be happy to volunteer trying an alpha) then let's roll with it.

**Comment by theKashey** [comment]
Ok, let me explain.

The biggest constraint of V3 is a literal unusability for lots of people. All these folks who use decorators, recompose, and quite happy about it - they are all off the boat.

V1 injects update code on the each-file-basic, thus allows it to perform point updates, and thus allows it to work with a complex code. But actually, it does not respect any complete code dependencies - module self-accept is a dangerous thing.

V3 performs the work much more gently, rendering from top to bottom, relying on `registered` classes to perform updates. This is the power, this is a weakness. It is just impossible to `register` everything, but you have to.

`Reconciler` is just another way to detect changes. 
The idea is simply - you have a rendered React Three, and you are updating a _single_ class. And you want only a single component to be updated (as V1 do).

Solution:
1. (hydrate)Traverse the existing tree, to store component type positions. Like - I know that Button was inside Form, inside Page12. This part actually borrows code from react-deep-force-update.
2. (hotReplacementRender)Traverse the new tree, to spot the changes to update proxies. I this part I have to explain in deep.

Reconsiler:
1. To make reconciler works RHL have to wrap all components with proxy. In real - not all, but all stateless, and all classes with injected "REGENERATE_METHOD". As long standin-proxy did not affect performance - we can do it.
2. Call render method on `real` component instance, obtained from hydrate.
3. Traverse the result, comparing component types.
3.1 In case of DOM node perform monkey patching (mergeInject) combining type from rendered child, and children from hydrated one.
3.2 In case of function compare the new and the old one
3.2.1 If types are equal - continue render
3.2.1 If types are equal from proxy point of view (`registed`) - continue render
3.2.1 If types are `swappable` - update proxy and continue render.

And `swappable` is the key of this solution.
In v3 `swappable` means that types have been exported from a same file and same name.
Here - it does mean - types were found in the same tree position, have the same display name and also some `similarity` in code.

`similarity` - full-text similarity, or 20% by Levenshtein distance (fuzzyCompare, up to O(n^2) ). For classes text similarity are tested on method level, allowing you add or remove one method, and change another. (has no effect on the `registered` types).

And there is a small trick here - reconsiler does not respect a possible change in props. It does not `render` the true react tree. It tries to re-render application as it was, to spot and swap changes.
If you are:
 - rendering a different component – it will don't enter that leaf
 - rendering component in a different order - it will swap types, and render them in the old order
 - rendering with different props, resulting a different nested tree - it will update as many proxies as it could, and the nested tree will be completely updated in any case.

So, this is not a real render, this is just another way to rehydrate hot updated code. 
It is still important to have a registered types, to perform major updates upon them, but now RHL can stand any decorator or function composition between these registered types.
Hooray?

File: packages/react-hot-loader/src/AppContainer.dev.js [hunk 6]
```diff
@@ -99,6 +112,7 @@ AppContainer.propTypes = {
   },
   errorReporter: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
   warnings: PropTypes.bool,
+  reconciler: PropTypes.bool,
```
Comment (neoziro) [inline packages/react-hot-loader/src/AppContainer.dev.js]: I think we should always activate reconciler. Why making it optional?

File: packages/react-hot-loader/src/AppContainer.dev.js [hunk 7]
```diff
@@ -99,6 +112,7 @@ AppContainer.propTypes = {
   },
   errorReporter: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
   warnings: PropTypes.bool,
```
Comment (neoziro) [inline packages/react-hot-loader/src/AppContainer.dev.js]: So we can drop this option isn't it?

**Comment by theKashey** [comment]
RE:
```js
// from
compose(one, two, three)(Component)
// to
compose(one, three)(Component)
```
Reconciler will do it's best, and update all the proxies. But react-dom will not stand a tree change.
WE could invent a way, to rehydrate the old instances into a new components, but I'll prefer not to hack React (more deeply).
